action.yml

```diff
@@ -254,17 +260,30 @@ runs:
       ;
       ~/.local/bin/ansible-test ${{ inputs.testing-type }}
       -v --color
-      --docker ${{ inputs.docker-image }}
       --coverage
       ${{
           inputs.testing-type == 'integration'
           && '--retry-on-error --continue-on-error --diff'
           || ''
       }}
       ${{
-          inputs.target-python-version
-          && format('--python {0}', inputs.target-python-version)
-          || ''
+          inputs.testing-type == 'integration'
+          && inputs.ansible-core-version != 'stable-2.9'
+          && inputs.ansible-core-version != 'stable-2.10'
+          && inputs.ansible-core-version != 'stable-2.11'
+          && inputs.target-python-version
+          && inputs.controller-python-version
+          && inputs.controller-python-version != inputs.target-python-version
+          && format('--controller docker:default,python={0} --target docker:{1},python={2}',
+                    inputs.controller-python-version,
+                    inputs.docker-image || 'default',
+                    inputs.target-python-version)
+          || inputs.target-python-version
+          && format('--docker {0} --python {1}',
+                    inputs.docker-image,
+                    inputs.target-python-version)
+          || format('--docker {0}',
+                    inputs.docker-image)
```

Comment:
This is a huge conditional and I don't understand the precedence from the first glance. It'll need to be improved by either moving this into a separate step or, at least, putting parentheses with indentation in more places.

Reply:
How would you like this being done in Python? Should the Python part compose the command in a previous step, and this step only run that composed command?

Reply:
I don't have a specific preference right now. We don't have to rewrite it in this PR, I was mostly voicing my vision for the future in general.

Reply:
I completely agree that this should be done in a nicer way (and Python code counts as nice) :-)